### DISCUSSION TITLE
Suggestion not to use / as identity separator if using IDs in URLs

### DISCUSSION THREAD

**Discussion by mcrio** [ravendb/ravendb#19167] [General]
Hi. This is just a suggestion to think twice about using the default identity separator `/` if there is any intention to use IDs as part of URLs. _Of course we would url-encode those, but there still might be issues._

It's annoying there are still proxies out there that by default transform `%2F` back to `/`, which will lead to HTTP 404 on your end.

I can't believe one of those is the Azure App Service rev-proxy. I'd expect Azure would conform to the URL RFC and not transform that `%2F`, given they are a top cloud provider.

Oren described a solution to [encrypt/decrypt the IDs](https://ayende.com/blog/191297-C/avoiding-exposing-identifier-details-to-your-users) but unless you want to hide the IDs, it seems like introducing more complexity due to the need to handle urls, dtos, input models..

Another solution would be to encode multiple times, but you never know how many proxies the request might hit or if they might change the behavior at some point.

In case of a green field project, I guess the easiest way is just to go with a different identity separator, one that is not a URL reserved character.

**Comment by mcrio** [comment]
But there seems to be a problem with server generated IDs.

If we want the server to append a suffix to the id, we have to end the id with `/`, and again we are stuck with the slash. So it has nothing to do with the client conventions related to the id separator.

Any workaround for this one?

**EDIT**: I see this works when we set the Client Identity Parts Separator at the database level. So setting it at client level only won't work.

**Reply by ppekrol** [reply]
Hi @mcrio 
As you noticed, the Client Identity Parts Separator set at Database level should resolve your issue, but if I may ask, why are you generating the IDs on the server-side? Client API uses a HiLo algorithm for identity generation (and takes into account the separator set), this does not suffice for your case?

**Reply by mcrio** [reply]
Hi @ppekrol 
Those are some user related log documents, generated as a side effect, which have the user id in the prefix, so at that point the quickest way to accomplish this was to let the server fill up the remaining part of the id.

Example:

IdentityPartsSeparator = '-'

PUT DOCUMENT:  `users-1-A-secretactionlog-`
PUT DOCUMENT:  `users-1-A-secretactionlog-`
PUT DOCUMENT:  `users-99-A-secretactionlog-`
PUT DOCUMENT:  `users-99-A-secretactionlog-`

**Reply by ayende** [reply]
Note that you can absolutely generate that on the client side. 
You can use the same hilo RavenDB is using internally.
